File: c++/0859.cpp
// Link: https://leetcode.com/problems/buddy-strings/description/
// Runtime: 0 ms
// Memory: 6.7 MB

class Solution 
{
public:
    bool buddyStrings(const std::string& s, const std::string& goal) 
    {
        if (s.size() != goal.size()) return false;

        int firstDifferentSymbolIndex = -1, secondDifferentSymbolIndex = -1;
        if (s == goal)
        {
            std::unordered_map<char, int> sCounter;
            for (const char& symbol : s)
            {
                sCounter[symbol]++;

                if (sCounter[symbol] > 1)
                    return true;
            }
        }
        else
        {
            for (int i = 0; i < static_cast<int>(s.size()); i++)
                if ((s[i] != goal[i]) && (firstDifferentSymbolIndex == -1))
                    firstDifferentSymbolIndex = i;
                else if ((s[i] != goal[i]) && (secondDifferentSymbolIndex == -1))
                    secondDifferentSymbolIndex = i;
                else if (s[i] != goal[i])
                    return false;
        }

        if (secondDifferentSymbolIndex == -1) return false;
        
        return s[firstDifferentSymbolIndex] == goal[secondDifferentSymbolIndex] &&
               goal[firstDifferentSymbolIndex] == s[secondDifferentSymbolIndex];
    }
};

File: python/0859.py
# Link: https://leetcode.com/problems/buddy-strings/description/
# Runtime: 51 ms
# Memory: 16.6 MB

class Solution:
    def buddyStrings(self, s: str, goal: str) -> bool:
        if len(s) != len(goal):
            return False
        
        first_different_symbol_index, second_different_symbol_index = -1, -1
        if s == goal:
            s_set = set()
            for symbol in s:
                if symbol in s_set:
                    return True
                s_set.add(symbol)
        else:
            for index, value in enumerate(s):
                if (s[index] != goal[index]) and (first_different_symbol_index == -1):
                    first_different_symbol_index = index
                elif (s[index] != goal[index]) and (second_different_symbol_index == -1):
                    second_different_symbol_index = index
                elif s[index] != goal[index]:
                    return False
        
        if second_different_symbol_index == -1:
            return False

        return s[first_different_symbol_index] == goal[second_different_symbol_index] and goal[first_different_symbol_index] == s[second_different_symbol_index]
